sdk/core/azure-core/inc/azure/core/websockets/websockets.hpp

```diff
@@ -0,0 +1,220 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+/**
+ * @file
+ * @brief Websockets functionality.
+ */
+
+#pragma once
+
+#include "azure/core/dll_import_export.hpp"
+#include "azure/core/exception.hpp"
+#include "azure/core/io/body_stream.hpp"
+#include "azure/core/url.hpp"
+
+#include <functional>
+#include <memory>
+#include <string>
+
+namespace Azure { namespace Core { namespace Websockets {
+
+  /**
+   * @brief Configuration for a websocket client.
+   *
+   */
+  struct WebsocketClientOptions
+  {
+  };
+
+  /**
+   * @brief The different types of websocket message.
+   *
+   * @details Text type contains UTF-8 encoded data. Interpretation of Binary type is left to the
+   * application.
+   *
+   */
+  enum class WebsocketMessageType
+  {
+    TextMessage,
+    BinaryMessage,
+    Close,
+    Ping,
+    Pong
+  };
+
+  /**
+   * @brief Template class for a websocket message.
+   *
+   */
+  template <class T> class WebsocketMessage {
+  protected:
+    Azure::Core::IO::BodyStream& m_bufferStream;
+
+    /**
+     * @brief Destroy the Websocket Message object
+     *
+     */
+    ~WebsocketMessage() = default;
+
+  public:
+    /**
+     * @brief Construct a new Websocket Message object
+     *
+     * @param type
+     * @param bufferStream
+     */
```

Comment:
Does it make sense for there to be other overloads of the WebSocketMessage constructor that take (for instance) a `std::string const&` which format a `TextMessage` message type? And what are the semantics of `Ping` which include a body?